### PR TITLE
Added initial unit tests for the runtime

### PR DESCRIPTION
**Pull request by austoonz**
*Description of changes:*

**Warning, this is a large PR!**

---

## Overview

Introduces unit testing framework for the PowerShell Lambda Runtime with coverage of runtime components.

### Testing Infrastructure

-   **Test execution framework** (`Invoke-Tests.ps1`) with coverage reporting
-   **Helper libraries** for HTTP API simulation, environment management, and assertions
-   **Testing standards documentation** (`tests/README.md`)

### Test Coverage

-   **Unit tests for 8 private runtime functions** covering success paths, error handling, and edge cases
-   **Module and build script tests**
-   **HTTP API simulation** for Lambda Runtime API testing

### Documentation & Build Updates

-   **Updated READMEs** with testing instructions
-   **Enhanced build script** for test builds
-   **Updated .gitignore** for test artifacts

## Usage

```powershell
# Run all tests with coverage
cd powershell-runtime && pwsh -NoProfile -Command "& './tests/Invoke-Tests.ps1' -Coverage"
```

This establishes the foundation for testing the PowerShell Lambda Runtime, supporting code quality and regression prevention.

---

The output from the above usage example (with some output truncated).

```powershell
PS> cd /dev/aws-lambda-powershell-runtime-unit-tests
PS> cd powershell-runtime/tests
PS> pwsh -NoProfile -Command "& ./Invoke-Tests.ps1 -Coverage -TestBuiltModule"
PowerShell Runtime Test Runner
===============================
Loading test requirements from: /dev/aws-lambda-powershell-runtime/powershell-runtime/test-requirements.psd1
Initializing test framework...
Found Pester version: 5.7.1
Pester imported
Test paths: /dev/aws-lambda-powershell-runtime/powershell-runtime/tests/unit
TestBuiltModule: True
Built module testing mode - building module for validation testing
Building PowerShell runtime module...
Module build completed
Module built - tests will use built module

Configuration:
  Test Type: All
  Test Mode: Built Module
  Build Module in Tests: False
  Test Source Files: False
  Coverage: True
  Parallel: False
  Output Format: Console
  CI Mode: False
  Detailed Output: False
  PowerShell Verbose: False
Code Coverage Configuration:
  Built Module: True
  Specific Path: False
  Test Source: False
  Files: 1
  Paths:
    - layers/runtimeLayer/modules/pwsh-runtime.psm1

Executing tests...
Configuring test containers with parameters: TestBuiltModule
Configured 1 test containers

Starting discovery in 10 files.
Discovery found 281 tests in 244ms.
Starting code coverage.
Running tests.
...
...
...
Tests completed in 7.83s
Tests Passed: 281, Failed: 0, Skipped: 0, Inconclusive: 0, NotRun: 0
Processing code coverage result.
Covered 87.93% / 80%. 174 analyzed Commands in 1 File.

Test Results Summary:
===================
Total Tests: 281
Passed: 281
Failed: 0
Skipped: 0
Duration: 00:00:07.8265606

Code Coverage: 87.93% (153/174 commands)

All tests completed
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by jnunn-aws** [comment]
Amazon Q is also good at organizing a mix of changes into logically categorized commits. 

image

File: powershell-runtime/tests/unit/Private/Invoke-FunctionHandler.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,710 @@
...
+    param($LambdaInput, $LambdaContext)
+    return "Function string response"
+}
+'@
+
+    Set-Content -Path (Join-Path $script:MockHandlersPath "function-handlers.ps1") -Value $script:FunctionScriptContent
+
+    # Create a test module for module handler testing
+    $script:TestModuleContent = @'
+function Invoke-TestModule {
```
Comment (jnunn-aws) [inline powershell-runtime/tests/unit/Private/Invoke-FunctionHandler.Tests.ps1:94]: Nit: Personally prefer to use a scriptblock or a separate ps1 file instead of a here string to keep syntax highlighting.

File: powershell-runtime/tests/unit/Private/Invoke-FunctionHandler.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,710 @@
...
+    $script:MockHandlersPath = Join-Path $script:TestFixturesPath "mock-handlers"
+    $script:TestModulesPath = Join-Path $script:TestFixturesPath "test-modules"
+
+    New-Item -ItemType Directory -Path $script:TestFixturesPath -Force | Out-Null
+    New-Item -ItemType Directory -Path $script:MockHandlersPath -Force | Out-Null
+    New-Item -ItemType Directory -Path $script:TestModulesPath -Force | Out-Null
+
+    # Create mock script handlers for testing
+    $script:SimpleScriptHandler = @'
+param($LambdaInput, $LambdaContext)
```
Comment (jnunn-aws) [inline powershell-runtime/tests/unit/Private/Invoke-FunctionHandler.Tests.ps1:42]: Nit: Personally prefer to use a scriptblock or a separate ps1 file instead of a here string to keep syntax highlighting.

File: powershell-runtime/tests/unit/Private/Set-HandlerEnvironmentVariables.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+            # Store original values for restoration
+            $script:OriginalRequestId = $env:AWS_LAMBDA_RUNTIME_AWS_REQUEST_ID
+        }
+
+        AfterEach {
+            # Restore original values
+            $env:AWS_LAMBDA_RUNTIME_AWS_REQUEST_ID = $script:OriginalRequestId
+        }
+
+        # It "Should require Headers parameter" {
```
Comment (jnunn-aws) [inline powershell-runtime/tests/unit/Private/Set-HandlerEnvironmentVariables.Tests.ps1:483]: Should not commit commented-out code.

**Review by jnunn-aws — APPROVED** [review]

**Comment by afroz429** [comment]
everything looks good. I have two observations.

When i run the unit tests on Windows, `20 ` tests are failing with similar error related to paths. Below is an example
```
[-] Get-Handler.When handler type is Script.Should detect script handler with .ps1 extension 32ms (31ms|0ms)
 Expected strings to be the same, but they were different.
 String lengths are both 21.
 Strings differ at index 9.
 Expected: '/var/task/handler.ps1'
 But was:  '/var/task\handler.ps1'
            ---------^
```

When i ran the unit tests in the clouddesktop, one test is failing.
```
[-] Set-HandlerEnvironmentVariables.When handling missing or invalid header values.Should handle empty string header values 21ms (19ms|2ms)
 Expected <empty>, but got $null.
 at $env:AWS_LAMBDA_RUNTIME_AWS_REQUEST_ID | Should -Be '', /local/home/afrozam/git/aws-lambda-powershell-runtime/powershell-runtime/tests/unit/Private/Set-HandlerEnvironmentVariables.Tests.ps1:271
 at <ScriptBlock>, /local/home/afrozam/git/aws-lambda-powershell-runtime/powershell-runtime/tests/unit/Private/Set-HandlerEnvironmentVariables.Tests.ps1:271
```
It's weird how environment variables are behaving in different environments when they have empty string values. I would expect below code to return `True` then `False`

In Windows PowerShell. (unexpected)
```

PS > # set environment variable to empty string
PS > $env:x = ''
PS > $env:x -eq ''
False
PS > $env:x -eq $null
True
```


In Windows, pwsh 7.5.2  (expected)

```
➜ # set environment variable to empty string

➜ $env:x = ''
➜ $env:x -eq ''
True
➜ $env:x -eq $null
False
```

on clouddesktop  (unexpected)

```
PS > $psversiontable

Name                           Value
----                           -----
PSVersion                      7.3.4
PSEdition                      Core
GitCommitId                    7.3.4
OS                             Linux 5.10.238-215.956.amzn2int.x86_64 #1 SMP Tue Jul 1 20:18:47 UTC 2025
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

PS > # set environment variable to empty string
PS > $env:x = ''
PS > $env:x -eq ''
False
PS > $env:x -eq $null
True
```

File: powershell-runtime/tests/unit/Private/Set-HandlerEnvironmentVariables.Tests.ps1 (hunk 3 above)
Comment (austoonz) [reply powershell-runtime/tests/unit/Private/Set-HandlerEnvironmentVariables.Tests.ps1:483]: Removed in a following change.

File: powershell-runtime/tests/Invoke-Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,886 @@
...
+
+function Exit-TestScript {
+    param (
+        $ExitCode,
+        $ExitMessage
+    )
+    $env:PSModulePath = $script:CurrentPSModulePath
+
+    # In CI mode, always exit the process
+    if ($CI) {
```
[the remark this answers is not in the document]
Comment (austoonz) [reply powershell-runtime/tests/Invoke-Tests.ps1:184]: This script has some updates and refactoring in upcoming PRs. Will take this into consideration.

**Review by afroz429 — APPROVED** [review]
